### PR TITLE
adds interpolate option in source resampling + adds source creation (…

### PR DESCRIPTION
**Pull request by NicoKiaru**
…3d sampler)

Hi @tischi, 

Do you think this PR is ok ? It's adding a function to create a source from another one but with a different voxel sampling.

Best, 

Nico

### CODE REVIEW

File: src/main/java/sc/fiji/bdvpg/sourceandconverter/importer/NewSourceAndConverterGetter.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            ) {
+        this.nx = nx;
+        this.ny = ny;
+        this.nz = nz;
+        this.at3D = at3D;
+        this.name = name;
+        this.imgfactory = imgfactory;
+    }
+
+    public NewSourceAndConverterGetter(
```
Comment (tischi) [inline src/main/java/sc/fiji/bdvpg/sourceandconverter/importer/NewSourceAndConverterGetter.java]: Could you give this class a more specific name? This is very generic 😄 at least I cannot guess from the name what it is doing.... 

In addition, is it maybe more a `Creator` or a `Factory` than a `Getter`? With `Getter` I associate that I get an object that exists already...

**Review by tischi — COMMENTED** [review]
And could you try to make whatever happens in `ResampledSource` easier to understand?
I guess `ResampledSource` is the place where the main stuff happens?
I think a way could be to split up the code in `getSource` into smaller functions and give those functions names that are very descriptive.

**Comment by NicoKiaru** [comment]
Come on, it's so easy to understand, no need for comments!

ok ok I'll comment a bit ;-)

Briefly (I know it's impossible to understand anything right now) : the idea is, if you have a complex source (warped several times, affine transformed, etc...) , to be able to 'resample' it as a normal 3d image with a certain voxel size. Then it can be exported as a new dataset or imageplus.

I'll put some comments and a demo of it.

**Comment by tischi** [comment]
> Briefly (I know it's impossible to understand anything right now) : the idea is, if you have a complex source (warped several times, affine transformed, etc...) , to be able to 'resample' it as a normal 3d image with a certain voxel size. Then it can be exported as a new dataset or imageplus.

Ok, does that just mean (i) to add another "resample view" on top or (ii) to make an actual RAM copy where all views/ converters are applied? If it is (ii) it may often not be possible due to limiting RAM, right?

File: src/main/java/sc/fiji/bdvpg/sourceandconverter/importer/NewSourceAndConverterGetter.java (hunk 1 above)
Comment (NicoKiaru) [reply src/main/java/sc/fiji/bdvpg/sourceandconverter/importer/NewSourceAndConverterGetter.java]: Changed to `EmptySourceAndConverterCreator` in https://github.com/bigdataviewer/bigdataviewer-playground/pull/97/commits/612860d6ac98ae3016a6b4353fc0fb049afa362a

**Comment by NicoKiaru** [comment]
> And could you try to make whatever happens in `ResampledSource` easier to understand?

I've added documentation everywhere and a demo `sc.fiji.bdvpg.bdv.sourceAndConverter.resample.ResamplingDemo`.

Check the demo to see how it could be useful (especially for source warped by BigWarp - I'll make a screencast one day) and I hope the documentation helps in `ResampledSource`. I do not believe it is easy to split `ResampledSource` in smaller parts.

**Comment by NicoKiaru** [comment]
Hi @tischi 

I'll work on trying to solve some issue that @martinschorb found today.

Are you ok with this PR ? I'd like to merge it if possible in the release.

If you think not, I can move the functions in another repo. Just let me know.

Best, 

Nico

**Comment by NicoKiaru** [comment]
Hey @tischi I will merge this PR because I need to move on and solve Martin issue. I need to have this PR because the version in the update site is too intricated with this branch. Sorry about this

**Comment by tischi** [comment]
No problem. I will only have time middle/end of next week again. I have to prepare a two day online course on image analysis...........

**Pull request merged**